tianshou/env/worker/subproc.py

```diff
@@ -0,0 +1,190 @@
+    np.int16: ctypes.c_int16,
+    np.int32: ctypes.c_int32,
+    np.int64: ctypes.c_int64,
+    np.float32: ctypes.c_float,
+    np.float64: ctypes.c_double,
+}
+
+
+class ShArray:
+    """Wrapper of multiprocessing Array"""
+
+    def __init__(self, dtype, shape):
+        self.arr = Array(_NP_TO_CT[dtype.type], int(np.prod(shape)))
+        self.dtype = dtype
+        self.shape = shape
+
+    def save(self, ndarray):
+        assert isinstance(ndarray, np.ndarray)
+        dst = self.arr.get_obj()
+        dst_np = np.frombuffer(dst, dtype=self.dtype).reshape(self.shape)
+        np.copyto(dst_np, ndarray)
+
+    def get(self):
+        return np.frombuffer(self.arr.get_obj(),
+                             dtype=self.dtype).reshape(self.shape)
+
+
+class SubprocEnvWorker(EnvWorker):
+    """Subprocess worker used in SubprocVectorEnv and ShmemVectorEnv."""
+
+    def __init__(self, env_fn: Callable[[], gym.Env], share_memory=False
+                 ) -> None:
+        super().__init__(env_fn)
+        self.parent_remote, self.child_remote = Pipe()
+        self.share_memory = share_memory
+        self.buffer = None
+        if self.share_memory:
+            dummy = env_fn()
+            obs_space = dummy.observation_space
+            dummy.close()
+            del dummy
+            self.buffer = SubprocEnvWorker._setup_buf(obs_space)
+        args = (self.parent_remote, self.child_remote,
+                CloudpickleWrapper(env_fn), self.buffer)
+        self.process = Process(target=_worker, args=args, daemon=True)
+        self.process.start()
+        self.child_remote.close()
+
+    def __getattr__(self, key: str):
+        self.parent_remote.send(['getattr', key])
+        return self.parent_remote.recv()
+
+    @staticmethod
+    def _setup_buf(space):
+        if isinstance(space, gym.spaces.Dict):
+            assert isinstance(space.spaces, OrderedDict)
+            buffer = {k: SubprocEnvWorker._setup_buf(v)
+                      for k, v in space.spaces.items()}
+        elif isinstance(space, gym.spaces.Tuple):
+            assert isinstance(space.spaces, tuple)
+            buffer = tuple([SubprocEnvWorker._setup_buf(t)
+                            for t in space.spaces])
+        else:
+            buffer = ShArray(space.dtype, space.shape)
+        return buffer
+
+    def _decode_obs(self, isNone):
+        def decode_obs(buffer):
+            if isinstance(buffer, ShArray):
+                return buffer.get()
+            elif isinstance(buffer, tuple):
+                return tuple([decode_obs(b) for b in buffer])
+            elif isinstance(buffer, dict):
+                return {k: decode_obs(v) for k, v in buffer.items()}
+            else:
+                raise NotImplementedError
+        return decode_obs(self.buffer)
+
+    def render(self, **kwargs) -> None:
+        self.parent_remote.send(['render', kwargs])
+        return self.parent_remote.recv()
+
+    def close(self) -> Any:
```

Comment:
Why remove self.closed? Shouldn't we assert whether the processes are closed before calling reset, step, etc, as discussed in #178, #166? Otherwise calling those methods while closed will raise BrokenPipe Error.

Reply:
I think people do not have to explicitly close environments. If they want to re-use them, just reset those environments; otherwise, just leave them there and garbage collection (``__del__``) will do the trick (makes sure the processes will be terminated.).